gitutils: add diffStrings, diffFiles, and use it in testament to compare expected vs gotten

timotheecour opened the pull request:
* add diffStrings, diffFiles (can be customized in future with options if needed)
* replaces all instances of `diff -uNdr` with an API call
* testament now uses diffStrings to compare expected vs gotten, making it easier to see what differs (in particular for whitespace issues, but not limited to that)
* expose `tempfiles.genTempPath`

## example
I added (and reverted) a temporary commit that added intentional changes in test "expected output", to show what it looks like in CI, see https://gist.github.com/timotheecour/515dde92cb29ce359a3f4b9546c247fd

## note: unrelated, preexisting bug in std/tempfiles
as shown in https://gist.github.com/timotheecour/515dde92cb29ce359a3f4b9546c247fd, it looks like `randomPathName` is buggy on windows: see https://github.com/nim-lang/Nim/issues/17898; 
this doesn't affect this PR though because I'm using a prefix diffStrings_a and diffStrings_b so there is no clash here (but it's still a bug that should be fixed in future work)

timotheecour committed:
gitutils: add diffStrings, diffFiles, and use it in testament to compare expected vs gotten

timotheecour committed:
_

timotheecour committed:
_

timotheecour committed:
fixup

xflywind commented on lib/std/tempfiles.nim:
```diff
@@ -82,13 +82,21 @@ template randomPathName(length: Natural): string =
     res[i] = state.sample(letters)
   res
 
+proc genTempPath*(prefix, suffix: string, dir = ""): string =
```
well,can't `createTempDir` use it too?

ringabout commented on lib/std/tempfiles.nim:
```diff
@@ -82,13 +82,21 @@ template randomPathName(length: Natural): string =
     res[i] = state.sample(letters)
   res
 
+proc genTempPath*(prefix, suffix: string, dir = ""): string =
```
Can't `createTempDir` use it too?

timotheecour replied:
good idea, done

ringabout commented on lib/std/tempfiles.nim:
```diff
@@ -82,13 +82,21 @@ template randomPathName(length: Natural): string =
+    dir = getTempDir()
+  result = dir / (prefix & randomPathName(nimTempPathLength) & suffix)
+
 proc createTempFile*(prefix, suffix: string, dir = ""): tuple[fd: File, path: string] =
```
You need to clean up the body of `createTempFile`.

remove

```
  var dir = dir
  if dir.len == 0:
    dir = getTempDir()
```

timotheecour replied:
i've now refactored via getTempDirImpl which simplifies code, but i can't remove this otherwise it keeps getting recomputed in the loop

timotheecour committed:
refactor with createTempDir

timotheecour committed:
cleanup

timotheecour committed:
refacotr

ringabout commented on lib/std/tempfiles.nim:
```diff
@@ -82,26 +82,34 @@ template randomPathName(length: Natural): string =
     res[i] = state.sample(letters)
   res
 
+proc getTempDirImpl(dir: string): string =
```
```suggestion
proc getTempDirImpl(dir: string): string {.inline.} =
```
Because it can be called for multiple times

ringabout replied:
yeah I mean `genTempDir` too

timotheecour replied:
done, but please read https://github.com/nim-lang/RFCs/issues/198#issuecomment-596968252, which shows that `{.inline.}` impacts whether cross-module inlining can occur (in absence of -flto); since that proc is internal, it shouldn't really matter, but i suppose it's cleaner for intent + for robustness against refactorings.

xflywind replied:
yeah I mean `genTempDir` too

timotheecour committed:
fixup

timotheecour committed:
fixup

timotheecour committed:
PRTEMP fake test spec changes to show effect of diffStrings

Araq commented:
We could revive `experimental/diff.nim` for this.

timotheecour committed:
add runnableExamples for experimental/diff + cross-reference with gitutils

timotheecour committed:
Revert "PRTEMP fake test spec changes to show effect of diffStrings"

This reverts commit 57dc8d642dce6c1127c98b7cbc9edbfe747d4047.

timotheecour commented:
PTAL
* added runnableExamples to experimental/diff and cross-linked both modules
* added notes in PR description showing how it looks like in CI when there's a diff (via a fake commit which inserted spec diffs which i've reverted)
* identified a pre-existing bug (https://github.com/nim-lang/Nim/issues/17898) which doesn't affect this PR

> We could revive experimental/diff.nim for this.

this can be done in future work and would be a 2 line fix once `experimental/diff` gains the required features for this:
* it doesn't have a human readable output yet, we could in future add `toPretty` to render it. 
* note also that `git diff` has more options (eg --word-diff, --color-moved etc) and honors a users' git config to customize how it's rendered (eg with colors etc)

Currently, a primary use case for `experimental/diff` over `git diff` is when shelling out is not an option (eg because it needs to be called at high frequency etc). This is not the case here.

timotheecour commented on testament/testament.nim:
```diff
@@ -307,7 +308,7 @@ proc addResult(r: var TResults, test: TTest, target: TTarget,
       maybeStyledEcho styleBright, expected, "\n"
```
showing `expected` is still useful (even with showing the diff) so users can copy paste it
(and showing the `given` is also useful for understanding the output)
